pkg/kube/secrets.go

```diff
@@ -165,36 +165,40 @@ func (r *secretsReader) ListByLocalObjectReferences(ctx context.Context, refs []
 	return secrets, nil
 }
 
-func (r *secretsReader) ListImagePullSecretsByPodSpec(ctx context.Context, spec corev1.PodSpec, ns string) ([]corev1.Secret, error) {
-	// try to fetch image pull secret from pod spec
-	secrets, err := r.ListByLocalObjectReferences(ctx, spec.ImagePullSecrets, ns)
-	if err != nil {
-		return nil, err
-	}
-
+func (r *secretsReader) getServiceAccountByPodSpec(ctx context.Context, spec corev1.PodSpec, ns string) (*corev1.ServiceAccount, error) {
 	serviceAccountName := spec.ServiceAccountName
 	if serviceAccountName == "" {
 		serviceAccountName = serviceAccountDefault
 	}
-	// try to fetch image pull secret from service account
-	var sa corev1.ServiceAccount
-	err = r.client.Get(ctx, client.ObjectKey{Name: serviceAccountName, Namespace: ns}, &sa)
+
+	sa := &corev1.ServiceAccount{}
+	err := r.client.Get(ctx, client.ObjectKey{Name: serviceAccountName, Namespace: ns}, sa)
 	if err != nil {
 		return nil, fmt.Errorf("getting service account by name: %s/%s: %w", ns, serviceAccountName, err)
 	}
-	serviceAccountSecrets, err := r.ListByLocalObjectReferences(ctx, sa.ImagePullSecrets, ns)
+	return sa, nil
+}
+
+func (r *secretsReader) ListImagePullSecretsByPodSpec(ctx context.Context, spec corev1.PodSpec, ns string) ([]corev1.Secret, error) {
+	imagePullSecrets := spec.ImagePullSecrets
+
+	sa, err := r.getServiceAccountByPodSpec(ctx, spec, ns)
+	if err != nil {
+		return nil, err
+	}
+	imagePullSecrets = append(imagePullSecrets, sa.ImagePullSecrets...)
+
+	secrets, err := r.ListByLocalObjectReferences(ctx, imagePullSecrets, ns)
 	if err != nil {
 		return nil, err
 	}
+
+	// TODO(erikgb): Why do we want to error out if the following happens? Is that what K8s do?
```

Comment:
TODO can be removed it is not actionable

Reply:
Could you elaborate, please? 🤔 I am not using image pull secrets, but I have the impression that K8s just ignores missing secrets referred to from imagePullSecrets. Is this wrong?

Reply:
not sure we want the same behaviour  , we can avoid deploying a job. to me its desire behaviour

Reply:
most likely for a user to define an Imagepullsecret if he is planning to use secrets and not other way around

Reply:
Ok, I'll remove the TODO.